packages/material/src/complex/table-array.control.tsx

```diff
@@ -41,117 +55,325 @@ export const tableArrayTester: RankedTester = rankWith(10, and(
+          }
+
+          const clone = _.clone(array);
+          clone.push(element);
+
+          return clone;
+        }
+      )
+    );
+  };
+
+  return (
+    <Toolbar hidden={true}>
+      <Grid container alignItems='center' justify='space-between'>
+        <Grid item>
+          <Typography type='title' className={labelClass}>{label}</Typography>
+        </Grid>
+        <Grid item hidden={{smUp: allErrors.length === 0}}>
+          <ValidationIcon id='tooltip-validation' errorMessages={allErrors}/>
+        </Grid>
+        <Grid item>
+          <Tooltip id='tooltip-add' title={`Add to ${labelObject.text}`} placement='bottom'>
+            <Button
+              fab
+              color='primary'
+              className={buttonClass}
+              aria-label={`Add to ${labelObject.text}`}
+              onClick={addItem}
+            >
+              <AddIcon/>
+            </Button>
+          </Tooltip>
+          <Tooltip title='Delete'>
+            <Button
+              fab
+              className={buttonClass}
+              aria-label={`Delete`}
+              disabled={numSelected === 0}
+              onClick={handleOpenConfirmDelete}
+            >
+              <DeleteIcon />
+            </Button>
+          </Tooltip>
+        </Grid>
+      </Grid>
+    </Toolbar>
+  );
+};
+const generateCells = (Cell, resolvedSchema, rowPath, cellErrors?) =>
+  Object.keys(resolvedSchema.properties)
+  .filter(prop => resolvedSchema.properties[prop].type !== 'array')
+  .map(prop => {
+    const cellPath = Paths.compose(rowPath, prop);
+    const props = {
+      cellProperty: prop,
+      resolvedSchema,
+      rowPath,
+      cellPath,
+      errors: cellErrors
+    };
+
+    return <Cell key={cellPath} {...props}/>;
+  });
+
+const ValidationIcon = ({id, errorMessages}) => (
+  <Tooltip
+    id={id}
+    title={errorMessages.map((e, idx) => <div key={`${id}_${idx}`}>{e}</div>)}
+  >
+    <Badge badgeContent={errorMessages.length}>
+      <ErrorOutlineIcon color='error'/>
+    </Badge>
+  </Tooltip>
+);
+const EmptyTable = () => (
+  <TableRow>
+    <TableCell>No data</TableCell>
+  </TableRow>
+);
+const TableHeaderCell = ({cellProperty}) =>
+  <TableCell >{cellProperty}</TableCell>;
+
+const TableWithContent = tableProps => {
+  const {data, path, resolvedSchema, childErrors,
+    handleClick, isSelected} = tableProps;
+
+  return data.map((child, index) => {
+    const childPath = Paths.compose(path, index + '');
+
+    return (
+      <TableRow
+        key={childPath}
+        hover
+        onClick={event => handleClick(event, child)}
+        selected={isSelected(child)}
+      >
+        <TableCell padding='checkbox'>
+          <Checkbox checked={isSelected(child)} />
+        </TableCell>
+        {generateCells(TableContentCell, resolvedSchema, childPath, childErrors)}
+      </TableRow>
+    );
+  });
+};
+const TableContentCell = ({rowPath, cellProperty, cellPath, errors, resolvedSchema}) => {
```

Comment:
It might be worthwhile to extract these helper components to a file of their own.